### PR TITLE
HTCONDOR-1577-string-token-iterator

### PR DESCRIPTION
**Pull request by GregThain**
So range-based for loops work with it, e.g.

for (auto &s: StringTokenIterator("JobUniverse MemoryUsage Foo")) {
	operation(s);
}

<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_utils/stl_string_utils.h [hunk 1]
```diff
@@ -139,22 +141,59 @@ void randomlyGenerateShortLivedPassword(std::string &str, int len);
...
+		next();
+		return *this;
+	}
+
+	bool operator!=(const StringTokenIterator &rhs) {
+		return (this->ixNext != rhs.ixNext) || (this->pastEnd != rhs.pastEnd);
+	}
+	bool operator==(const StringTokenIterator &rhs) {
+		return this->ixNext == rhs.ixNext && this->pastEnd == rhs.pastEnd;
+	}
```
Comment (JaimeFrey) [inline src/condor_utils/stl_string_utils.h:189]: Shouldn't these also check whether str and delims match?

Comment (GregThain) [reply src/condor_utils/stl_string_utils.h:189]: It is (intentionally) undefined behavior (tm) to try to compare two iterators to different container instances, so I htink this is ok.

Comment (JaimeFrey) [reply src/condor_utils/stl_string_utils.h:189]: StringTokenIterator is acting as both a container and the iterator over that container, which makes the semantics goofy.

File: src/condor_utils/stl_string_utils.h [hunk 2]
```diff
@@ -139,22 +141,59 @@ void randomlyGenerateShortLivedPassword(std::string &str, int len);
...
+
 protected:
 	const char * str;   // The string we are tokenizing. it's not a copy, caller must make sure it continues to exist.
 	const char * delims;
 	int ixNext;
 	std::string current;
+	bool pastEnd;
```
Comment (JaimeFrey) [inline src/condor_utils/stl_string_utils.h:196]: I know size is not really an issue for these objects, but putting the pastEnd before current does make it smaller.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
The commits should be redone with the 10.4.0 cleanup ticket number (HTCONDOR-1577).

**Comment by timtheisen** [comment]
Let's now make more work for ourselves. I'll accept the old ticket number.

File: src/condor_utils/stl_string_utils.h (hunk 1 above)
Comment (GregThain) [reply src/condor_utils/stl_string_utils.h:189]: If you like, we can make the iterator proper a separate class.

**Review by JaimeFrey — APPROVED** [review]
LGTM.
Maybe we should revisit the ticket-per-release for the cleanup code.

**Pull request merged**